Send via SMTP support for aliases

sahilph opened the pull request:
This PR add ability to send via SMTP for aliases. With this emails can be sent directly from aliases via SMTP without the need for creation of reverse-alias.

If enabled, Every alias will have a unique password (generated client side) for SMTP access. Username will be same as alias.

This adds a global option in Settings to enable this feature, without enabling this option, SMTP access will remain disabled for the user.

Note: Even after enabling this option in settings, you would still need to enable SMTP individually per alias. This will also disable receiving via reverse-alias for that particular alias. i.e. Sender's original address will be displayed in "From:".

SMTP server address should ideally be the same where your app is hosted (but this depends on your configuration)
SMTP port: 465
This would need SSL cert/key file as SMTP would be over (implicit) SSL.

This should resolve #679 

The screenshots below will make it more clear on how it would operate:

- First enable "SMTP for Aliases" from Settings Page:

![image](https://user-images.githubusercontent.com/812110/161362225-30c1d345-54e1-4f09-bfb2-c342443c27a2.png)

- Then on the Aliases page, select "More" on an alias and a new toggle, to enable SMTP, will have appeared below "Pin this alias":

![image](https://user-images.githubusercontent.com/812110/161362304-8a1bb917-58cb-4f11-9ceb-41fa88f2bc4f.png)

- After toggling it, SMTP credential will be generated for that alias:

![image](https://user-images.githubusercontent.com/812110/161362409-b2eca965-a219-425f-8970-d5a04ac8978a.png)

sahilph committed:
Create initial SMTP handler

sahilph committed:
Add helper functions to SMTPCredentials class

sahilph committed:
Modify Front-end for SMTP support

sahilph committed:
Fix smtp-for-aliases-section checkbox issue

sahilph committed:
Change script type to module 

Also change way of importing nanoid

sahilph committed:
Improve SMTP password logic

sahilph committed:
Add __repr__ to SMTP Credentials class

sahilph committed:
Add decode to auth_data in SMTP_handler

sahilph committed:
Revert "Add auto-generated migrations"

This reverts commit 42c4c806ec0f3c151f7892799a1fcf9a21878930.

sahilph committed:
Add auto-generated migrations

sahilph committed:
Change Default port to 465 for SMTP_handler

sahilph committed:
Format code with: poetry run black

sahilph committed:
Fix email logging

sahilph committed:
Re-enable is_SMTP in EmailLog

sahilph committed:
Fix filtering for SMTP enabled aliases

sahilph committed:
Revert "Add auto-generated migrations"

This reverts commit ba2a047e52f8bd7b5600466e8df020a6d6320d8a.

sahilph committed:
Add auto-generated migrations

sahilph committed:
Add SMTP variables to example.env

sahilph committed:
Add about SMTP handler to README

developStorm commented:
Great job! Thanks @sahilph. I think this is definitely a valuable feature. The only point I currently find somewhat awkward is that each alias would get their own credentials. My thought for this feature is that users should be able to connect to the SMTP server using a generic credential, and the SMTP server would automatically detect which specific alias should be used to send the email by looking at the address the user is replying to, and other information that available. This way, users would only have to add their credentials once to Gmail's send-as feature or local email client, and reply to emails sent to any alias with the generic credentials.

sahilph commented:
@developStorm Thanks for you feedback.

One of the main reasons I developed this was to use Gmail "Send mail as:" feature. And It requires users to add "From:", and username and password for every alias. 
So even if there was a single generic credential, users will have to manually add the alias in Gmail anyways. So I thought of adding separate credentials, which would make the authentication handling easier.

Also consider one more scenario. Suppose the aliases are getting generated on-the-fly and I want to give a email to website say `non-existent-alias@mycustom.domain`. 
Now `non-existent-alias@mycustom.domain` is not the alias which I ever send email from, but for some reason the website needs replying from that email, even if there was generic credential then I would have to again add that alias in Gmail's "Send mail as:" anyways.

In the current scenario, since we can enable SMTP per-alias-basis, a reverse-alias will automatically get generated and replying will be easier as I can just reply via reverse-alias.

Also I'm not sure how to authenticate and validate, if there was only a single generic credential present. If you have any ideas please let me know and I can implement it.

developStorm commented:
@sahilph The idea is that the SMTP server does not need to honor the `From` specified in "Send mail as:" feature. So we can just put a dummy `From` there and let the SMTP server figrue out which `From` (alias) should be used based on the mail content that is being sent. How to determine right alias to use is a problem though, I'm thinking that we can use the recipent's address but I'm not sure if that's necessarily accurate. Any thoughts on this?

sahilph commented:
@developStorm While it would be very convenient to have such a feature, I can't really think of a way of implementing this without reading `From:` field :sweat_smile:.

developStorm commented:
@sahilph We just have to guess ;-) For example, if the incoming email to SMTP server looks like:

```
MAIL FROM: dummy@simplelogin.co
RCPT TO: user@example.com
```

Then now SMTP server can guess: "hmm the authenticated user has a recent interaction with `user@example.com` on alias `alias1@simplelogin.co`, so we should change the From to `alias1@simplelogin.co`". SMTP server then delivers the reply on behalf of `alias1@simplelogin.co` to `user@example.com`.

sahilph commented:
@developStorm I don't think we should rely on guesswork while sending emails. 

Also what about sending to someone who we have never sent before ?

developStorm commented:
@sahilph Yeah guess is not a perfect idea, but I think it's worse to have user add every single alias they have, when they have let's say 50+ aliases and need to be able to reply as all those aliases. 

For new recipients, user can manually add them on SL portal before they proceed, as what they have to do now to obtain reply address.

sahilph commented:
> Yeah guess is not a perfect idea, but I think it's worse to have user add every single alias they have, when they have let's say 50+ aliases and need to be able to reply as all those aliases.

User don't have to add every single alias. They only need to add the ones they usually send from. For other aliases those who dont have SMTP enabled, sending via reverse-alias is still possible 

> For new recipients, user can manually add them on SL portal before they proceed, as what they have to do now to obtain reply address.

Then what's the point of SMTP? users can do that already do that using reverse-alias.

The whole point of SMTP access (and this PR) is that, if users want to send emails from the aliases they most often send from, there would be no need of creating reverse-alias. Users can just the emails using their aliases via SMTP using their favorite email client. No need to manually login to SL portal and create reverse aliases everytime.

For other non-frequently used aliases, use the reverse-alias method as that would still work.

TL;DR: For sending via most-frequently used alias -> Use SMTP ; otherwise -> use the regular-reverse alias method.

sahilph commented:
@developStorm 
Check out  #679 as to what I mean. I was facing similar issues.
Here are a few comments highlighting the issue:
https://github.com/simple-login/app/discussions/679#discussioncomment-1622066
https://github.com/simple-login/app/discussions/679#discussioncomment-1622905

nguyenkims commented on SMTP_handler.py:
```diff
@@ -0,0 +1,520 @@
+            LOG.e("Incorrect Format for Credentials")
+            return self.fail_nothandled(status.E501)
+
+        username = (auth_data.login).decode("utf-8")
```
we can remove the parenthesis around `auth_data.login` right?

sahilph replied:
I have removed Unnecessary parenthesis.

nguyenkims commented on SMTP_handler.py:
```diff
@@ -0,0 +1,520 @@
+            return self.fail_nothandled(status.E501)
+
+        username = (auth_data.login).decode("utf-8")
+        password = (auth_data.password).decode("utf-8")
```
⏫

nguyenkims reviewed (commented):
The PR looks really good! I haven't run it though, it'd be great if you can include how to test this new service in the CONTRBUTING file.

I left some comments, lots of them are rather nits though.

sahilph committed:
Add command to generate self-signed certs in README

sahilph committed:
change SMTP_password to lowercase.

sahilph committed:
Suffix PATH to key/cert vars  to be explicit

sahilph committed:
add unique=True in alias_id declaration

sahilph committed:
Change set_password part

- Also change comment to " override PasswordOracle.password" to be more explicit
- Add one more check in alias.py view while saving password.

sahilph committed:
Revert "ReFormat index and setting htmls using Pycharm"

This reverts commit aec499e6bfdfe4612008064c056da3dbc205b05e.

sahilph committed:
ReFormat index and setting htmls using Pycharm

- Also change JS variable to camelCase in index.html

sahilph committed:
Add auto-generated migrations

sahilph committed:
Revert "Add auto-generated migrations"

This reverts commit 25a56ec7fc910b2bb9b6ab819b1cd13a67192f2c.

sahilph committed:
Get alias domain by get_email_domain_part in SMTP_handler

Also:
- remove should_ignore_bounce
- Remove region comment
- move "user.disabled" higher to avoid the unnecessary contact creation
- Remove unnecessary parenthesis

sahilph committed:
Add "Send a test email via SMTP" in CONTRIBUTING

sahilph commented on README.md:
```diff
@@ -489,6 +489,32 @@ docker run -d \
+
+For sending via SMTP  you need to make sure you have the required SSL certificate and key at following, as SMTP requires SSL.
+
+- `$(pwd)/smtp_ssl_cert.pem` for SSL Certificate file,
```
I have added a command to generate self-signed certificates.

sahilph commented on README.md:
```diff
@@ -489,6 +489,32 @@ docker run -d \
+    -v $(pwd)/dkim.pub.key:/dkim.pub.key \
+    -v $(pwd)/smtp_ssl_cert.pem:/smtp_ssl_cert.pem
+    -v $(pwd)/smtp_ssl_privkey.key:/smtp_ssl_privkey.key
+    -p 465:465 \
```
Not sure how do it via Postfix.
Just out of curiosity, how would protecting this port via postfix prevent attacks ?

nguyenkims replied:
Postfix can block requests from bots quite efficiently. And if we add spam scanners like SpamAssassin or rspamd to Postfix, we can be quite sure that requests that reach the SMTP handler is legit.

sahilph replied:
Oh ok. Gotcha. 
But the authentication should be the first thing that happens, so does Postfix allows for a programmatic way of handling authentication ? I couldn't seem to find any good way to handle auth in postfix's docs.

One more way to handle bots is that we could use Fail2ban for this post, they allow for creation of custom filters quite easily. Then later we are sending it to postfix anyways, there it can handle SPAM.

So Basically the flow can be like:
Port 465 (Protected by Fail2Ban) -> SMTP_Handler.py -> Postfix (SPAM handling) -> Final Recipient

nguyenkims replied:
> But the authentication should be the first thing that happens, so does Postfix allows for a programmatic way of handling authentication ? I couldn't seem to find any good way to handle auth in postfix's docs.

That's a good question and we're actually looking for the answer as well :).

nguyenkims replied:
I'm not familiar with fail2ban unfortunately. I think we still need Postfix to handle TLS, grey listing, spam scanning, slow connections that take up workers and all other edge cases due to the email protocol. One way to have Postfix is:
1) postfix listening on port 587/465 that requires SMTP AUTH
2) postfix talks to Python to verify whether the authentication is correct. This can be tricky as as postfix only support sasl authentication using dovecot and cyrus.
3) If authentication is correct, Postfix relays to SMTP handler

It'd be great if we can modify the PR to use this pattern.

sahilph replied:
> I'm not familiar with fail2ban unfortunately. 

Fail2ban basically detects the authentication failures and bans the corresponding IP. the ban time can be set. This is very useful to prevent brute-force attacks. You can check that out here: https://www.fail2ban.org/


> 2. postfix talks to Python to verify whether the authentication is correct. This can be tricky as as postfix only support sasl authentication using dovecot and cyrus.

Yes, I as well, couldn't figure out how to make postfix talk to python. 


> It'd be great if we can modify the PR to use this pattern.

That would certainly be great. We would somehow need to figure out the "2." part.

sahilph commented on app/api/views/alias.py:
```diff
@@ -318,6 +319,23 @@ def update_alias(alias_id):
+        enable_SMTP = data.get("enable_SMTP")
+        if enable_SMTP:
+            if "SMTP_password" in data:
+                SMTP_password = data.get("SMTP_password")
```
I have changed this.

sahilph commented on app/api/views/alias.py:
```diff
@@ -318,6 +319,23 @@ def update_alias(alias_id):
+                SMTP_password = data.get("SMTP_password")
+                if (
+                    SMTP_password and len(SMTP_password) != 21
+                ):  # 21 is default password length set for nanoid.
```
This implementation Nanoid works client-side. 
So, yes, I think we would need to add it to other Simple Login clients as well.

nguyenkims replied:
I wonder if the client can send password longer than 21 characters or it must be 21 characters?

sahilph replied:
It can be be of any length. I am checking with 21 as it was the default length mentioned in their docs which they say is secure enough.

nguyenkims replied:
In this case should we replace the check `!=` by `<`?

nguyenkims replied:
ok let's keep the 21 value then.

sahilph commented on app/config.py:
```diff
@@ -444,3 +444,7 @@ def setup_nameservers():
 NAMESERVERS = setup_nameservers()
+
+# For SMTP support
+SMTP_SSL_CERT_FILE = os.environ.get("SMTP_SSL_CERT_FILE") or "/smtp_ssl_cert.pem"
```
> Can we add `PATH` suffix to be explicit, so `SMTP_SSL_CERT_FILEPATH` instead of `SMTP_SSL_CERT_FILE`?

I have changed this.

sahilph replied:
> And as existing installations don't have the certificates and keys file, should this new service disabled unless the env variables are set?

The service won't start at all, if the files cert/key files do not exist and thus would remain disabled.

sahilph commented on app/models.py:
```diff
@@ -2557,6 +2575,45 @@ class AliasMailbox(Base, ModelMixin):
 
+class SMTPCredentials(Base, ModelMixin, PasswordOracle):
+    __tablename__ = "SMTP_credentials"
+    __table_args__ = (sa.UniqueConstraint("alias_id", name="uq_alias"),)
```
I have added `unique=True` in `alias_id` declaration

nguyenkims replied:
Nits: With `unique=True` added, we can remove this `__table_args__` I think

sahilph replied:
I have removed the `__table_args__`

sahilph commented on app/models.py:
```diff
@@ -2557,6 +2575,45 @@ class AliasMailbox(Base, ModelMixin):
+        sa.ForeignKey(Alias.id, ondelete="cascade"), nullable=False, index=True
+    )
+
+    # override, SMTP password should not be null
```
I have added `override PasswordOracle.password`

sahilph commented on app/models.py:
```diff
@@ -2557,6 +2575,45 @@ class AliasMailbox(Base, ModelMixin):
+            pass_oracle = PasswordOracle()
+            pass_oracle.set_password(password)
+            password = pass_oracle.password
+            smtp_cred: SMTPCredentials = super(SMTPCredentials, cls).create(
```
I have changed this part.

sahilph commented on email_handler.py:
```diff
@@ -927,28 +932,32 @@ def forward_email_to_mailbox(
-        add_or_replace_header(msg, "Reply-To", new_reply_to_header)
-        LOG.d("Reply-To header, new:%s, old:%s", new_reply_to_header, reply_to_header)
+    # Don't Change Headers when SMTP is enabled for alias
+    if not is_SMTP_enabled_for_alias:
```
Won't VERP handle this ? SPF record are checked against the envelope sender and not the `From:` header, right ?

The reason I added this was to prevent confusion and threading issue in the mail client.

Say user is using Gmail's "Send Mail as:" and send a email to a contact. 
Then, if this is not there, then while getting reply, the `From:` will be be a reverse-alias. It will cause the email to not be in the thread and might create confusion for the user.

nguyenkims replied:
You're right, the email will pass SPF. The email doesn't have envelope and header alignment and will fail DMARC though. 

The threading is handled by the headers "Message-ID", "In-Reply-To" and "References". 

To handle the case where reverse alias and contact address are mixed, I think we can handle the case the envelope.rcpt_to is a reverse alias in SMTP handler. In this case we can consider that the same as envelope.rcpt_to is the contact address. Let me know if that makes sense :).

sahilph replied:
> You're right, the email will pass SPF. The email doesn't have envelope and header alignment and will fail DMARC though.

Yes, for DMARC to pass the original DKIM signature should be preserved. 

--
> The threading is handled by the headers "Message-ID", "In-Reply-To" and "References".

Oh ok ok
--

> To handle the case where reverse alias and contact address are mixed, I think we can handle the case the envelope.rcpt_to is a reverse alias in SMTP handler. In this case we can consider that the same as envelope.rcpt_to is the contact address. Let me know if that makes sense :).

 I did not get this part. Can you please explain this ?

nguyenkims replied:
> I did not get this part. Can you please explain this ?

Currently SMTP handler receives an email from an alias to a contact and send the email to the contact.

What I meant is in addition to that flow, SMTP handler can also handle the case user sends an email from an alias to a reverse alias. So when SMTP handler receives an email from an alias to a reverse alias, it extracts the contact address from the reverse alias and continue the normal process.

nguyenkims replied:
> So you are saying that, the SMTP handler should handle this case as well, right?

Yep.

> Also, in this scenario, the reverse-alias, should belong to the alias, correct ?

Yep.

This case can happen when:

1) user sends an email to a contact via SMTP (so no reverse alias)
2) contact replies, the response will be sent from the reverse alias
3) user replies back, the reply will be sent from the alias to the reverse alias -> it's where SMTP handler is called so we need to handle this case.

sahilph replied:
> 2. contact replies, the response will be sent from the reverse alias
 
I certainly believe this would cause confusion to the users. 
We can just forward the email as-is without changing the `From:` to reverse-alias.

I get that you are concerned that it will fail DMARC, but I think all we need to do is preserve the original DKIM signature, then add our DKIM signature (which is already being done), the DMARC will be passed.
 I guess that is how the pure forwarding services (like ImprovMX, Cloudflare Email Routing, etc) do it.

We can also look into adding the ARC set headers if the emails are still ending up in spam.

nguyenkims replied:
As I mentioned earlier, the email will fail dmarc alignement so we can't keep the original header. We **cannot** keep original header at the moment. This code part needs to be reverted in order for the PR to be merged.

ARC is another topic that should probably be addressed in another PR.

sahilph replied:
I have reverted this part.

sahilph commented on static/package-lock.json:
```diff
@@ -1,8 +1,209 @@
   "name": "simplelogin",
   "version": "1.0.0",
-  "lockfileVersion": 1,
+  "lockfileVersion": 2,
```
Yes, it was probably due to different npm version. I used v16, which version do you use ?

sahilph commented on templates/dashboard/index.html:
```diff
@@ -190,6 +190,13 @@
               Pinned Aliases
             </option>
+
+          {%  if current_user.enable_SMTP_aliases %}
```
I have re-formatted using Pycharm

sahilph commented on templates/dashboard/index.html:
```diff
@@ -277,6 +284,13 @@
                       title="This alias is pinned"></span>
               {% endif %}
 
+            {%  if current_user.enable_SMTP_aliases %}
```
I have re-formatted using Pycharm

sahilph commented on templates/dashboard/index.html:
```diff
@@ -588,4 +619,81 @@
+          let newValue = !oldValue;
+
+          if(newValue) {
+            var smtp_generated_password = nanoid(21);
```
I have fixed this

sahilph commented on SMTP_handler.py:
```diff
@@ -0,0 +1,520 @@
+    """
+    website_email = rcpt_to
+
+    # region Create variables For Logging Purpose
```
Yes, Probably. 
I haved removed the region comment.

sahilph commented on SMTP_handler.py:
```diff
@@ -0,0 +1,520 @@
+        LOG.e("Alias: %s isn't known", alias_address)
+        return False, status.E503
+
+    alias_domain = alias_address[alias_address.find("@") + 1 :]
```
I have changed this.

sahilph commented on SMTP_handler.py:
```diff
@@ -0,0 +1,520 @@
+
+    contact = Contact.get_by(website_email=website_email)
+    if not contact:
+        LOG.w(f"No contact with {website_email} as website email")
```
No, not necessarily, the contact might exist, if user has sent via SMTP before.

nguyenkims replied:
What I meant is this is a rather normal case so we shouldn't use logging warning level here. info or debug level is more suitable.

sahilph replied:
I have changed this to debug level.

sahilph commented on SMTP_handler.py:
```diff
@@ -0,0 +1,520 @@
+        LOG.w(f"No contact with {website_email} as website email")
+    try:
+        LOG.d("Create or get contact for website_email:%s", website_email)
+        contact = get_or_create_contact("", website_email, alias, msg)
```
I didn't understand this. Can you please explain a bit more ?

nguyenkims replied:
I mean the current `get_or_create_contact` handles a lot of cases that aren't necessary for this usage. I think it's better to have a smaller method, which is a stripped down version of the `get_or_create_contact` to be used here.

sahilph replied:
I have created a similar stripped down version `get_or_create_contact_for_SMTP_phase` please check if its fine.

sahilph commented on SMTP_handler.py:
```diff
@@ -0,0 +1,520 @@
+        alias = Alias.get_by(email=alias_address)
+        if not alias:
+            LOG.i("Alias %s was deleted in the meantime", alias_address)
+            if should_ignore_bounce(envelope.mail_from):
```
Yes, that makes sense. I have removed this.

sahilph commented on SMTP_handler.py:
```diff
@@ -0,0 +1,520 @@
+    mailbox = Mailbox.get_by(id=alias.mailbox_id)
+    # endregion
+
+    if user.disabled:
```
I have moved it higher.

sahilph commented:
> The PR looks really good! I haven't run it though, it'd be great if you can include how to test this new service in the CONTRBUTING file.
> 
> I left some comments, lots of them are rather nits though.

@nguyenkims Thanks for the review

I have added how to test this in the CONTRIBUTING file. I have also done some fixes as per you comments. Please check.

sahilph commented on SMTP_handler.py:
```diff
@@ -0,0 +1,520 @@
+                return [(False, status.E515)]
+
+    user = alias.user
+    mailbox = Mailbox.get_by(id=alias.mailbox_id)
```
Well, if user can have multiple inboxes, it won't be relevant to include mailbox here. 
I added it only for logging as I thought Email Log requires it.  But it doesn't, so it can be removed. What are your thoughts. ?

nguyenkims replied:
The `mailbox_id` is needed by EmailLog. In this reply phase, this is the mailbox (or a mailbox's authorized address) that sends the email.

sahilph replied:
But for SMTP phase, the `envelope.mail_from`, will be the alias itself.

So how to get the relevant mailbox in case of multiple mailboxes ?

I think it just just be removed as we are already setting `is_SMTP=True`

sahilph committed:
Re-format again with PyCharm

- Changed PyCharm's reformat settings to match the repo

sahilph committed:
Redo Change JS vars to camelcase

- Had to do this again after commit revert

sahilph commented:
Also, It seems like PyCharm does not have a consistent reformatting settings across different versions for html files. 
I have now guessed and set the reformat settings and they now seem to match with this repo.

sahilph commented:
Hello @nguyenkims , I was wondering whether you got time to look into this?

nguyenkims reviewed (commented):
Hey I tested the PR and it works well locally. I just left some comments.

nguyenkims commented:
@sahilph we have been actually discussing a lot about the PR in our team :). The 

> Hello @nguyenkims , I was wondering whether you got time to look into this?

Yes we have extensively discussed about the PR during our team meeting :). Before enabling this SMTP server, we need to make sure it is scalable and has a good anti abuse system.

sahilph commented:
>  Yes we have extensively discussed about the PR during our team meeting :). Before enabling this SMTP server, we need to make sure it is scalable and has a good anti abuse system.

That is good to hear. :)

Scalability wise, I think it should be fine as you already use aiosmtpd, and I believe it works good enough.

For anti abuse system as I mentioned in the above comments, we could use Fail2ban for this port, they allow for creation of custom filters quite easily. Then later we are sending it to postfix anyways, there it can handle SPAM.

So Basically the flow can be like:
Port 465 (Protected by Fail2Ban) -> SMTP_Handler.py -> Postfix (SPAM handling) -> Final Recipient

nguyenkims commented on app/models.py:
```diff
@@ -2557,6 +2575,45 @@ class AliasMailbox(Base, ModelMixin):
+            )
+            del pass_oracle
+        else:
+            return None
```
@sahilph  In case you haven't seen this comment.

nguyenkims replied:
Yeah return None is enough.

nguyenkims reviewed (commented):
I just answered some questions.

sahilph reviewed (commented):
@nguyenkims I have replied to some of the comments.

nguyenkims reviewed (commented):
Hey just left a thought. I think we can merge the PR and it can be used as-is for the self hosting. For SimpleLogin.io version, I think we need to figure out a way to have Postfix protecting this new service.

sahilph commented:
> I think we can merge the PR and it can be used as-is for the self hosting. 

Oh that's nice.

> For SimpleLogin.io version, I think we need to figure out a way to have Postfix protecting this new service.

Yes, if we could figure out a way to make postfix talk to python, that would be great as that is the only thing preventing this from happening.

sahilph commented:
Also, I have replied to the comments.

nguyenkims reviewed (commented):
I just left 2 comments.

nguyenkims commented on SMTP_handler.py:
```diff
@@ -0,0 +1,515 @@
+        LOG.e("%s domain isn't known", alias)
+        return False, status.E503
+
+    contact = Contact.get_by(website_email=website_email)
```
As mentioned in another comment, I think we can also handle the case where website_email is a reverse alias here. This should be relatively easy IMO.

sahilph replied:
I have added some code to handle this case.

nguyenkims reviewed (commented):
I realized that I haven't done the follow up on the discussion on SMTP_handler file.

sahilph committed:
Remove `__table_args__` when `unique=True` in models.py

sahilph committed:
Revert keeping original header in email_handler.py

sahilph committed:
Handle sending to reverse alias in SMTP_handler

- Also create a stripped down function `get_or_create_contact_for_SMTP_phase` 
- Also change the log to debug instead of warning

sahilph committed:
Add auto-generated migrations

sahilph reviewed (commented):
@nguyenkims I have made some changes and left some comments.

acasajus commented on app/api/views/alias.py:
```diff
@@ -318,6 +319,25 @@ def update_alias(alias_id):
+        else:
+            SMTPCredentials.delete_by_alias_id(alias_id=alias.id)
+        alias.enable_SMTP = enable_SMTP
+        changed = True
```
Instead of generating the password in the client it'd be better to generate it server-side and show it once to the client on successful activation:

 - No need to add extra dependencies in python and in js (ie. nanoid)
 - No need to have js code in the front
 - No strange 21 character validation
 - Other clients have no need to add nanoid dependencies.
 
Just generate a random secret, store it in the db and in the next page display show the credentials warning the user that the credentials won't be able to be displayed any more times.

sahilph replied:
@nguyenkims Is this really needed at this stage ?

nguyenkims replied:
@sahilph  generally I think it's better to generate secret on server as we can't control client environment. Adding this change should be quite simple right as the server generates the password and send this back to the client to display?

acasajus commented:
Hey, nice PR. But as a general comment, there aren't any tests added. I would suggest to add some:

 - Creating the SMTPCredentials does what it's supposed to and that passwords can be validated
 - The SMTP handler is able to receive connections, authenticate/fail, and deliver the messages to where it's expected.

sahilph commented:
@acasajus Thanks for the feedback. 
 Yes, I understand that there are no tests, and I can certainly write some if needed, but I believe, ideally,  tests should be written by someone else, because if I also write tests for this, it will be highly biased. (I certainly don't want my one code to break now do I ? :grin: )

nguyenkims commented:
@sahilph @acasajus Having tests for every PR is indeed a good idea. We actually set a limit in terms of code coverage under which a PR won't pass the CI. And tests should be written by the PR author :). It happens several times that writing tests help discover cases that we don't realize during the development. We are not extreme in terms of test coverage though but there should be enough tests to cover the most frequent cases.

Robin-Sch commented:
any update? I really need this asap

EvilAaron commented:
Yeah, I am following this with great interest as well.

I purchased a year of SL and love much about the service and have brought 2 of my domain on board. 

After using for a couple of weeks though and mucking with the reverse alias system and clogging my contacts up with extra redundant emails in their profiles, this is the one part that feels clunky. 

Really hoping this is solved with this above proposed solution. If no way to "send as" SL aliases from email client arrives in a year when renewal arrives, I would have to consider pulling my domains and finding a service for email support and relegate SL to just throwaway emails on their domains.

I am hoping a way will be found, as I love SL so far and would love for it to fit.

therealmrmuffin commented:
I'm not sure i get the point of this. Seen as SimpleLogin Is not indeed really meant to be a transactional email service. But if there are other use cases for then, then i would say. Why not.

nephalemsec commented:
What exactly is holding this back? This has been sitting in a completed state for so long, how can we get it over the line?

edsimpsons83 commented:
Agreed, disappointing to see such a valuable PR just sitting here.

nguyenkims commented:
@nephalemsec @edsimpsons83 the PR isn't complete. We are asking for adding tests into it (among other things) and haven't received any news yet. And please keep in mind that this PR is rather for self hosting. To make it available on SL.io, we need to adapt the infra and work on an anti-abuse system first.

nephalemsec commented:
> @nephalemsec @edsimpsons83 the PR isn't complete. We asked for adding tests into it (among other things) and haven't received any news yet. And please keep in mind that this PR is rather for self hosting. To make it available on SL.io, we need to adapt the infra and work on an anti-abuse system first.

Understood. As @sahilph has stated he believes there would be a bias with the PR author making unit tests is there a plan for SL team to assist?

nguyenkims commented:
@nephalemsec providing unit tests should actually be part of the PR and it is up to the author to write tests.

therealmrmuffin commented:
> I'm not sure i get the point of this function, seen as SimpleLogin is indeed not _meant_ to be a transactional email service. But if there are other use cases for this function, then i would say, "why not".

Sadly i didn't get a reply from someone explaining the purpose of this function to me. :(

sugarfunk commented:
> > I'm not sure i get the point of this function, seen as SimpleLogin is indeed not _meant_ to be a transactional email service. But if there are other use cases for this function, then i would say, "why not".
> 
> Sadly i didn't get a reply from someone explaining the purpose of this function to me. :(

To be able to send from an alias without needing to log into the SL interface comes to mind.

jasonhe54 commented:
So, is this PR just dead now after 6 months? Is there any intent on supporting this feature? Would be nice to actually be able to use SL as a way to send emails from an alias similar to the way iCloud Plus currently lets you do it.

Haulien commented:
It's a shame this PR has been left to die. Time to look for a better alternative to SL.

EvilAaron commented:
> It's a shame this PR has been left to die. Time to look for a better alternative to SL.

Any luck on hunt for alternatives. Have my own domains, thinking of alternatives to move to once current year sub at SL ends...

Haulien commented:
> > It's a shame this PR has been left to die. Time to look for a better alternative to SL.
> 
> Any luck on hunt for alternatives. Have my own domains, thinking of alternatives to move to once current year sub at SL ends...

I changed to Fastmail. Handles all the alias stuff I could ever want.

HyperCriSiS commented:
> > > It's a shame this PR has been left to die. Time to look for a better alternative to SL.
> > 
> > 
> > Any luck on hunt for alternatives. Have my own domains, thinking of alternatives to move to once current year sub at SL ends...
> 
> I changed to Fastmail. Handles all the alias stuff I could ever want.

Was it possible to transfer all this easily without significant interruption?
It was so much work with all those aliases and I also still have severe issues with the sieve filters in Proton Mail not working properly.

Haulien commented:
> Was it possible to transfer all this easily without significant interruption?

yeah you setup a wildcard and it's barely any work. Just keep in mind you can't disable aliases in the same way. It's basically making a rule to send anything to the alias to the trash. So depending on how many you have disabled, that could take a bit to setup.

trikotret commented:
> > Was it possible to transfer all this easily without significant interruption?
> 
> yeah you setup a wildcard and it's barely any work. Just keep in mind you can't disable aliases in the same way. It's basically making a rule to send anything to the alias to the trash. So depending on how many you have disabled, that could take a bit to setup.

Actually you can block an alias. The sender will get a notification that alias doesn't exist. You just create any alias you want to block and then disable it.

HyperCriSiS commented:
Why is this not merged??

nguyenkims commented:
This MR was created by a community member and isn't complete. We have asked several times but haven't heard back from the original author:

https://github.com/simple-login/app/pull/869#issuecomment-1264693168

sahilph committed:
Remove nanoid requirement

sahilph committed:
Update SMTP_handler.py

sahilph committed:
fix enable/disable SMTP api call

sahilph committed:
Update example.env

sahilph committed:
Update SMTP_handler.py

No need to send message manually. Correct message is automatically sent.

sahilph committed:
Une existing function to create

sahilph committed:
Fix SMTP Creation in Templates

sahilph committed:
Add Config Vars

sahilph committed:
Add SMTP Handler Tests

sahilph committed:
Add background running in SMTP Handler

sahilph committed:
Adapt SMTP test

sahilph committed:
Add migrations

sahilph committed:
Add SMTP auth helper

sahilph committed:
Update the env vars

sahilph committed:
Adapt tests for nginx smtp proxy

sahilph committed:
Remove unused imports

sahilph committed:
Update README.md

sahilph commented:
Based on recent comments by various users, It seems that this PR is still very popular.

I have updated the code for the current master branch. I have also added tests.

As @nguyenkims has previously [mentioned](https://github.com/simple-login/app/pull/869#discussion_r841040198), it will  be not safe to directly expose `SMTP_handler` to the internet as it does not have any anti-abuse system. But, I still couldn't find a way for postfix to programmatically handle authentication. However, Nginx now supports SMTP proxying. Its already being used to [protect the api](https://github.com/simple-login/app/pull/869#discussion_r841040198), might as well use it to protect `SMTP_handler` as well. This should solve the anti-abuse issue. 
Hence I have updated the code to use Nginx as SMTP proxy and added the required helper functions. The request flow for SMTP will be:
Internet -> Nginx (handle auth) -> `SMTP_Handler`

To summarize:
- Code has been updated for latest master branch
- Tests have been added.
- Nginx acts as SMTP proxy and protects `SMTP_handler`

sahilph committed:
Merge branch 'master' into SMTP_for_aliases

sahilph committed:
Fix and Remove unused imports from alias.py

sahilph committed:
Remove obsolete note

sahilph committed:
Update SMTP_handler.py

sahilph committed:
Update models.py

sahilph committed:
Change Testing Port

sahilph committed:
Merge branch 'master' into SMTP_for_aliases

sahilph commented:
The is one other change that I forgot to add in my [previous comment](https://github.com/simple-login/app/pull/869#issuecomment-2928918571).

The password for SMTP will now get generated on server side, as per the @acasajus [previous suggestion](https://github.com/simple-login/app/pull/869#discussion_r852897910).

obadz commented:
This is amazing, that's exactly the feature I requested here: https://www.reddit.com/r/Simplelogin/comments/1kso1qj/ability_to_send_from_an_alias_via_smtp/ !

sahilph committed:
Merge branch 'master' into SMTP_for_aliases

sahilph committed:
Merge branch 'master' into SMTP_for_aliases

obadz commented:
Hi @nguyenkims, do you require any more changes before this PR can be merged? Cheers.

sahilph committed:
Merge branch 'master' into SMTP_for_aliases

sahilph committed:
Merge branch 'master' into SMTP_for_aliases

sahilph committed:
Merge branch 'master' into SMTP_for_aliases

sahilph commented:
> > The PR looks really good! I haven't run it though, it'd be great if you can include how to test this new service in the CONTRBUTING file.
> 
> Hi @nguyenkims, 6 months ago you left a very positive comment on this PR. I think @sahilph has addressed all the points you mentioned but hasn't gotten the required reviews/approvals needed to merge it. How can we get someone's attention on this PR which is a rather exciting feature?

I am wondering the same. Not sure why they are not responding to this PR, they seem to be active on other issues/PRs

sahilph committed:
Merge branch 'master' into SMTP_for_aliases

balthazar commented:
Fervent paid customer of simplelogin also badly missing this feature

HyperCriSiS commented:
> This is indeed very disappointing, especially since @nguyenkims left a very positive comment on this PR just couple of months back.
> 
> > @acasajus:  since it would break many of the anti-abuse measures that we have set up in place
> 
> This reason listed does not seem like a actual reason for closing the PR. Whatever anti abuse measures you have, I am sure that they can be incorporated into this, but instead of working on that, the PR was closed, which leads me to believe that there was some other real reason for closing the PR.
> 
> Don't take this this the wrong way, but it feels like ever since Proton acquired Simple Login, it has gone all corporate and don't really care much about what the community/users want... 😞

I also have this feeling. There is barely any development. But not only here, like at Proton general. They are just fixing bugs and integrate rather uninteresting things.

Whats the thing now with this Pull???

DNCD commented:
> This is indeed very disappointing, especially since @nguyenkims left a very positive comment on this PR just couple of months back.
> 
> @acasajus: since it would break many of the anti-abuse measures that we have set up in place
> This reason listed does not seem like a actual reason for closing the PR. Whatever anti abuse measures you have, I am sure that they can be incorporated into this, but instead of working on that, the PR was closed, which leads me to believe that there was some other real reason for closing the PR.
> 
> Don't take this this the wrong way, but it feels like ever since Proton acquired Simple Login, it has gone all corporate and don't really care much about what the community/users want... 😞

Hey @sahilph ,

I completely agree with you. I’m left feeling that the PR was closed without a clear explanation, which makes it seem like there might be an agenda beyond the stated anti‑abuse concerns. Rather than leaving the community in the dark, it would be helpful to know exactly which safeguards the change would break and what would be required to bring the contribution back in line with those protections.

stevenwashere commented:
Interestingly, proton mail has added smtp support which is awesome. But why not simple login?

I use custom domains through simple login because they do not count towards my limit but let me reply from there which is great for support situations. 

On proton each custom domain alias counts towards your max number even though it's all the same mailbox and is my own domain.

Maybe we can get this reconsidered since another part of their team is able to do it? I know that was not a open source implementation but to close this because they do not want to figure out a security aspect when another part of the org with more restrictions has recently opened up SMTP feels bad.

ppaslan commented:
Very disappointing. I should probably start to look for alternatives.

psycho-neon commented:
Why was this closed without even considering the review? is been ignored by the team.

sahilph commented:
>  is been ignored by the team.

Unfortunately, yes, it seems to be the case. 😞
